app/services/course/assessment/question/programming_codaveri/r/r_package_service.rb

```diff
@@ -102,13 +104,19 @@ def extract_supporting_solution_files
   # @param [Pathname] pathname The pathname of the file.
   # @param [String] content The content of the file.
   def extract_supporting_file(filename, content)
-    supporting_solution_object = default_codaveri_data_file_template
+    supporting_file_object = default_codaveri_data_file_template
 
-    supporting_solution_object[:type] = 'internal' # 'external' s3 upload not yet implemented by codaveri
-    supporting_solution_object[:path] = filename.to_s
-    supporting_solution_object[:content] = content
+    supporting_file_object[:type] = 'internal' # 'external' s3 upload not yet implemented by codaveri
+    supporting_file_object[:path] = filename.to_s
+    if content.force_encoding('UTF-8').valid_encoding?
+      supporting_file_object[:content] = content
+      supporting_file_object[:encoding] = 'utf8'
+    else
+      supporting_file_object[:content] = Base64.strict_encode64(content)
+      supporting_file_object[:encoding] = 'base64'
+    end
 
-    @data_files.append(supporting_solution_object)
+    @data_files.append(supporting_file_object)
   end
```

Comment:
Nit: time to consider moving this function out to a common file, given that this is language agnostic, and we'll be having C# and Javascript soon. Okay to wait till when we add those languages.

Reply:
I had the same thought, but still thinking over the code structure after doing that. We will definitely do so when we add the C# and JavaScript Codaveri services.